presto-main/src/main/java/com/facebook/presto/operator/repartition/MapBlockEncodingBuffers.java

```diff
@@ -0,0 +1,334 @@
+        return INSTANCE_SIZE +
+                getPositionsRetainedSizeInBytes() +
+                sizeOf(offsets) +
+                sizeOf(offsetsBuffer) +
+                sizeOf(offsetsCopy) +
+                getNullsBufferRetainedSizeInBytes() +
+                keyBuffers.getRetainedSizeInBytes() +
+                valueBuffers.getRetainedSizeInBytes() +
+                sizeOf(hashTablesBuffer);
+    }
+
+    @Override
+    public long getSerializedSizeInBytes()
+    {
+        return NAME.length() + SIZE_OF_INT +            // length prefixed encoding name
+                columnarMap.getKeyType().getTypeSignature().toString().length() + SIZE_OF_INT + // length prefixed type string
+                keyBuffers.getSerializedSizeInBytes() +    // nested key block
+                valueBuffers.getSerializedSizeInBytes() +  // nested value block
+                SIZE_OF_INT +                           // hash tables size
+                hashTableBufferIndex +                  // hash tables
+                SIZE_OF_INT +                           // positionCount
+                offsetsBufferIndex + SIZE_OF_INT +      // offsets buffer.
+                getNullsBufferSerializedSizeInBytes();  // nulls
+    }
+
+    private void populateNestedPositions()
+    {
+        // Reset nested level positions before checking positionCount. Failing to do so may result in elementsBuffers having stale values when positionCount is 0.
+        keyBuffers.resetPositions();
+        valueBuffers.resetPositions();
+
+        if (positionCount == 0) {
+            return;
+        }
+
+        offsets = ensureCapacity(offsets, positionCount + 1);
+
+        int[] positions = getPositions();
+
+        for (int i = 0; i < positionCount; i++) {
+            int position = positions[i];
+            int beginOffset = columnarMap.getOffset(position);
+            int endOffset = columnarMap.getOffset(position + 1);
+            int currentRowSize = endOffset - beginOffset;
+
+            offsets[i + 1] = offsets[i] + currentRowSize;
+
+            if (currentRowSize > 0) {
+                keyBuffers.appendPositionRange(beginOffset, currentRowSize);
+                valueBuffers.appendPositionRange(beginOffset, currentRowSize);
+            }
+        }
+    }
+
+    private void appendOffsets()
+    {
+        offsetsBuffer = ensureCapacity(offsetsBuffer, offsetsBufferIndex + batchSize * ARRAY_INT_INDEX_SCALE, LARGE, true);
+
+        int baseOffset = lastOffset - offsets[positionsOffset];
+        for (int i = positionsOffset; i < positionsOffset + batchSize; i++) {
+            offsetsBufferIndex = setIntUnchecked(offsetsBuffer, offsetsBufferIndex, offsets[i + 1] + baseOffset);
+        }
+        lastOffset = offsets[positionsOffset + batchSize] + baseOffset;
+    }
+
+    private void appendHashTables()
+    {
+        // MergingPageOutput may build hash tables for some of the smallers blocks. But if there're some blocks
```

Comment:
Curious if this can also be optimized (separately). Sounds like it's unnecessary for `MergingPageOutput` to eagerly build hash table. 

Also, is it possible to have mixed "map blocks with hash tables" and "map blocks without hash tables" sending to `MapBlockEncodingBuffers`? -- In this case, is it possible to have both types of single map blocks (with and without hash tables) in **one single output partition**?

Reply:
> Curious if this can also be optimized (separately). Sounds like it's unnecessary for MergingPageOutput to eagerly build hash table.

Great questions.  Yes we can definitely make MergingPageOutput not eagerly build the hash tables.

 > Also, is it possible to have mixed "map blocks with hash tables" and "map blocks without hash tables" sending to MapBlockEncodingBuffers?

Yes it is possible to have mixed "map blocks with hash tables" and "map blocks without hash tables" sending to MapBlockEncodingBuffers. This is because MergingPageOutput only merge small pages and build hash tables eagerly. So the blocks being merged have hash tables while original large blocks don't have hash tables. In such case we don't need to send the hash tables at all. This was the reason why we had the `noHashTables` field to identify if there're any hash tables or not. 

 > -- In this case, is it possible to have both types of single map blocks (with and without hash tables) in one single output partition?

I am not quite sure if I understand this. Did you mean SingleMapBlock? I don't think PartitionedOutputOperator would receive such a block, right?

Reply:
@yingsu00 

> I am not quite sure if I understand this. Did you mean SingleMapBlock? I don't think PartitionedOutputOperator would receive such a block, right?

No I mean what will happen when mixed "map blocks with hash tables" and "map blocks without hash tables" sending to MapBlockEncodingBuffers. And you already explained in the previous comment. Thanks!